### PR TITLE
add array destructuring to AQL

### PR DESCRIPTION
**Pull request by jsteemann**
This allows using a nicer syntax for parsing the members of an array into separate AQL variables.
Recursive destructuring is also supported

Example query that will demonstrate the new functionality:

    LET a = ["foo", "bar", "baz", [1, 2, 3], false, "meow"]
    LET [foo, bar, baz, [sub1, , sub3], , meow] = a
    RETURN { foo, bar, baz, sub1, sub3, meow }

This will return

    [
      {
	"foo" : "foo",
	"bar" : "bar",
	"baz" : "baz",
	"sub1" : 1,
	"sub3" : 3,
	"meow" : "meow"
      }
    ]

### CODE REVIEW

**Comment by dsonet** [comment]
Very useful, would like to see this feature.

**Comment by jsteemann** [comment]
It seems that it works already, but it is obviously missing tests.

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2806/

**Review by Simran-B — CHANGES_REQUESTED** [review]
Added suggestions to improve the docs. Spreading and destructuring also need to be explained in the regular documentation - shall I do the required changes in this PR?

File: Documentation/Books/Manual/ReleaseNotes/NewFeatures35.md [hunk 1]
```diff
@@ -51,10 +51,135 @@ For MMFiles this rule does not apply.
...
+    LET values = ['the', 'quick', 'foxx', 'jumps', 'over', 'the', 'dog'] 
+    RETURN CONCAT(values[0], values[1], values[2], values[3], values[4], values[5], values[6])
+
+The spread operator can also be used on input objects. In this case it will add all the
+attributes of the input object into the target object. 
+
+Consider the following example query:
+    
+    LET name = { first: "John", middle: "J", last: "Doe" }
+    RETURN { id: "1234", age: 39, ...obj, active: true }
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply Documentation/Books/Manual/ReleaseNotes/NewFeatures35.md]: Ok, thanks for the hint. I have used your suggestion in the NewFeatures35.md file.

**Comment by jsteemann** [comment]
@Simran-B : please feel free to adjust the manual as you see fit!

**Comment by Simran-B** [comment]
Object spreading doesn't behave as I expected. In JavaScript:

```js
console.log(JSON.stringify( {
  foo: 1, bar: 2,
  ...{baz: 3, foo:4, new: 5, nested: {sub: 6} },
  ...{new: 7, nested: {sub: 8} }
} ));
// { "foo": 4, "bar": 2, "baz": 3, "new": 7, "nested": { "sub": 8} }
```

In AQL I got:

```js
RETURN {
  foo: 1, bar: 2,
  ...{baz: 3, foo:4, new: 5, nested: {sub: 6} },
  ...{new: 7, nested: {sub: 8} }
}
// { "foo": 1, "bar": 2, "baz": 3, "new": 5, "nested": { "sub": 6 } }
```

Looks like the first value wins? That also raised the question whether object spreading is equivalent to MERGE() or rather MERGE_RECURSIVE(), but currently it seems to be neither.

**Comment by jsteemann** [comment]
@Simran-B : AQL does not support overwriting existing attributes. The exceptions here may be MERGE and MERGE_RECURSIVE. However, any other places keep the first value of each attribute and don't attribute it.

**Comment by Simran-B** [comment]
@jsteemann I know that literals like `{a: 1, a: 2}` lead to `{ a: 1 }`, which is different to the behavior of JavaScript, but acceptable. If `...{ }` does that however, then that seems really counter-intuitive. I would bet that many JS users don't know by heart what you end up with if you specify the same key in an object literal twice, but if they know the object spread operator then I'm sure they know that the value of the key specified last wins.

The current documentation is also wrong because it states that it would be the same as `MERGE()`, which it is not. It is rather like MERGE_RECURSIVE(), but with the arguments reversed (I think?)

Would it be difficult to change the behavior? Could the expressions be reversed internally so that "first key wins" doesn't get in the way?

**Pull request closed without merging**